### ISSUE TITLE
SceneBuilder needs an extensible UI, an extension API to grow and a reactive Gluon

### ISSUE THREAD

**Issue by treilhes** [gluonhq/scenebuilder#288]
Hello Gluon, Everyone,

After taking some time looking at the code of SceneBuilder the least we can say is this soft does not respect the DRY principle and everything from the UI to the features are set in stone. It is still a wonderfull software to build UI fast. Between the dinosaur QT and the frankenstein Electron on the fat client side, between native, Xamarin and others on the mobile side, javafx can become a leading technology but it isn't. Why? 

I think Java strong point lies in his large open source community, it gave us developement tools like Eclipse, IntelliJ and others which can rival paid IDE like Visual Studio due to the massive amount of plugins you can get on the net. The success of VSCode is in the same vein. Actualy, SceneBuilder is cut from this community and seems to be stuck at the same place Oracle put it 8 years ago. Opening SceneBuilder to community plugins/extensions should be the number one priority to fill the gap with other IDE but instead, issues and pull requests are stacking without feedback... :sob: 

Don't misunderstand me, i'm not blaming Gluon, i understand they are already doing a great job maintaining the project. Involving yourself into an open source side project of openjfx has a cost, gluon's business is not centered on the scenebuilder and the covid context is not helping either.

I plan to give some free time to this project but working to create a rejected/not reviewed pull request is not realy worth it.
As a starting point to the discussion, can you validate or comment the following points:

- Does SceneBuilder need to refactor his code to adhere to the DRY principle?
- Does SceneBuilder need to expose an API dedicated to extensions/plugins (Menu/UI/Task/...)? 
- Can other framework be allowed into SceneBuilder? Dependency injection? OSGI? RxJavafx? other?
- Which kind of reactivity can we expect about issues or pull request from Gluon?
- Something else?

PS: English is not my main language so be lenient please...

**Comment by palexdev** [comment]
Scene Builder is in maintenance mode, they just make sure the software works with the latest Java and JavaFX releases.
However, in my opinion they misunderstood how open-source software works. It doesn't mean "maintained by the community", it means "the community can contribute" which are different sentences. Also, there are actually some PRs but why don't they get merged, not even commented some of them
I love JavaFX and I love Scene Builder and that's why I'm worried.
Gluon... cut that shitty GraalVM thingy who the heck cares about that, put some real effort on Scene Builder and JavaFX because if you don't do that JavaFX will soon die.
I mean, already now it is not a beautiful situation, just search on the internet something like "Is JavaFX dead?" and all you will find are people who answer "yes", "use electron", "javafx is good but it's 'old' "

**Comment by georgemoralis** [comment]
started effords to modularize scenebuilder and update to latest . You can take a look on my fork on 

https://github.com/codebbgr/scenebuilder

**Comment by palexdev** [comment]
> started effords to modularize scenebuilder and update to latest . You can take a look on my fork on 
> (modules branch)
> https://github.com/codebbgr/scenebuilder
> 

Starred! Thanks

**Comment by georgemoralis** [comment]
just a drop, scenebuilder works in jdk15 and it's fully modular (JPMS)

**Comment by treilhes** [comment]
Hi @georgemoralis , @palexdev,

Glad to see i'm not alone on this thread. I've also started to work on a scenebuilder refactoring. My main motive is to allow developers to easily create extensions to scenebuilder the same way you create custom controls. 
@georgemoralis, It seems we didn't go the same way but i'm realy interested in your modularization work. Feel free to take whatever you feel usefull on my branch. I hope we can merge our branches soon

If you want to take a look https://github.com/treilhes/scenebuilder/tree/feature/docking-system

**Comment by georgemoralis** [comment]
hmm very nice , but why spring ?  I would choose a more lightweight DI  like 
https://github.com/SupernautApp/SupernautFX


I will take a look on your code @treilhes  , hope we can work together for a brighter future of scenebuilder

**Comment by palexdev** [comment]
@georgemoralis why did you remove gradle?

**Comment by treilhes** [comment]
@georgemoralis Why Spring? To be honest i've used Spring because i'm fluent with it. I did not made any pros/cons table. So your question made me think a bit more about this choice. After some deliberation with me and myself i've decided to stick with it for the following reasons:
- I know how to use it 
- Spring is also widely used/known in the Java world
- Performance is a concern but not a priority as SceneBuilder only need to handle one user
- Spring can be much more than only DI

The major cons that could make me change to another DI framework would be the fact that Spring support JPMS by using automatic module name which prevent the use of jlink and jpackage

Another important point, i don't plan (at least for now) to create a permanent fork. I hope Gluon will give some feedback at some point and will merge this refactoring into their master branch. So i've kept all features included by Gluon and made the less changes possible to the build system to keep whatever devops pipeline they made working.

Spirit of Gluon I summon you, rise from the dead and give us your opinion...

**Comment by georgemoralis** [comment]
@palexdev i hate gradle

**Comment by georgemoralis** [comment]
@treilhes  i doubt gluon (hope i am wrong) is interesting to work on scenebuilder. It's a good product but has a lot of bugs that needs to be sorted. Also most of the pull requests get stalled in gluon without a response.
My company plans to give some resources to scenebuilder developing and keep it separate from gluon branch , hopefully we can provide more updates in the near future

**Comment by johanvos** [comment]
@georgemoralis We are very interested to keep working on SceneBuilder. However, if there are no companies that want to pay us for maintenance and development, we can't do too much. We are making sure it keeps working with the latest JavaFX releases, and we keep "maintaining" and distributing it. 
I am hopeful we can increase our commitment, but tbh, SceneBuilder has much more takers than givers :(

Having said that, we really should reply to PR's, I am sorry if they are stalled and blocked on us. I will have a look at the list.

**Comment by treilhes** [comment]
Hello @johanvos ,
Is it a joke? Yes, it must be joke....
When you see the words "My company plans to give some resources to scenebuilder" it did take you 19 minutes to give an answer. But the main joke is that you didn't even give a tiny comment about the subject of this thread.
Are you for real?

**Comment by johanvos** [comment]
I started looking into your PR, and I asked @abhinayagarwal to look as well, because we were triggered by the notification mails. 
As how I would feel? Well, when I use a product that I don't pay for, and that is being maintained by somehow who is not paid for it, I would be thankful. 
It is not a joke, and yes, you are rude. Not just a little bit.

**Comment by treilhes** [comment]
Hello again @johanvos,

It seems there is a misunderstanding here, i don't care about my PR concerning binding expression it only took me 2 days of work. My main issue is my current development. I've spent roughtly 10 days refactoring scenebuilder source code and i think i will spend twice this amount in future month to achieve the goal explained at the start of this thread. More or less a 10K € gift to Gluon. What i don't want is creating a pull request that will be refused. So to prevent that i created this issue to get some guidance from Gluon about what i can or can't do with the source code. Without any answers from Gluon i still started to work taking the risk to throw this work to the toilet (to be honest i also need this dev so it won't be lost).

When you write:

> Well, when I use a product that I don't pay for, and that is being maintained by somehow who is not paid for it, I would be thankful.

I understand it fully and i'm thankfull (just read my first post in this thread and you will see that i am) but you also need to understand that i want to give Gluon free labor and to achieve this in the most profitable ways for you and me i need some answers to my questions. What i called a "joke" in my previous post was the fact that you replied to a post without replying to the subject of this thread. A little bit like if i was waiting for someone for days and when i meet him, he talks to the person next to me for 2min then run to somewhere else leaving me with an half open mouth (dunno if it's figurative enought).

So please i beg for your forgiveness, can you read the beginning of this thread?

**Comment by georgemoralis** [comment]
@treilhes  i am interesting for you "gift" if you provide me with a valid pull request for my fork 👍

**Comment by Blacklands** [comment]
@georgemoralis I was just reading the comments on this issue, and then I went looking for your fork. It doesn't seem to exist - did you delete/private it? Just asking out of curiosity!

**Comment by treilhes** [comment]
@abhinayagarwal, @johanvos 
Just a simple question, do you have any existing end to end test ? What kind of tech is involved? can you share some details abut them if any?
